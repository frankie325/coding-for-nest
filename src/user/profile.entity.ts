import { Column, Entity, OneToOne, PrimaryGeneratedColumn } from 'typeorm';
import { User } from './user.entity';

@Entity()
export class Profile {
  @PrimaryGeneratedColumn()
  id: number;
  @Column()
  gender: number;
  @Column()
  photo: string;
  @Column()
  address: string;

  // 创建一对一的关系
  @OneToOne(() => User)
  user: User;
}
