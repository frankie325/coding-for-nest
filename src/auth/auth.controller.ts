import {
  Body,
  Controller,
  Post,
  UseGuards,
  Request as NestRequest,
  UseInterceptors,
  ClassSerializerInterceptor,
} from '@nestjs/common';
import { AuthService } from './auth.service';
import { User } from '../user/user.entity';
import { SigninUserDto } from './dto/signin-user.dto';
import { AuthGuard } from '@nestjs/passport';
import { Request } from 'express';
// import { SerializeInterceptor } from 'src/interceptors/serialize/serialize.interceptor';

@Controller('auth')
@UseInterceptors(ClassSerializerInterceptor) // 使用拦截器：序列化排除password字段
export class AuthController {
  constructor(private authService: AuthService) {}

  // AuthGuard('local')将会使用Passport的本地策略进行验证
  // 如果验证成功，用户信息将会被注入到请求对象中
  // 如果验证失败，将会抛出401 Unauthorized异常
  // @UseGuards(AuthGuard('local'))
  @Post('/signin')
  signin(@Body() dto: SigninUserDto, @NestRequest() req: Request) {
    // console.log('🚀 ~ AuthController ~ signin ~ req.user:', req.user);
    const { username, password } = dto;
    return this.authService.signin(username, password);
  }

  // @UseGuards(AuthGuard('jwt'))
  @Post('/signup')
  // @UseInterceptors(SerializeInterceptor)
  signup(@Body() dto: SigninUserDto) {
    const { username, password } = dto;
    return this.authService.signup(username, password);
  }
}
