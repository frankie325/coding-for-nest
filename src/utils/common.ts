import { Role } from '../roles/roles.entity';
import { Logs } from '../logs/logs.entity';
import { User } from '../user/user.entity';
import { Menu } from '../menus/menu.entity';

// 根据path映射为casl的subject
export const getEntities = (path: string) => {
  const map = {
    '/users': User,
    '/logs': Logs,
    '/roles': Role,
    '/menus': Menu,
    // '/auth': 'Auth',
  };

  for (let i = 0; i < Object.keys(map).length; i++) {
    const key = Object.keys(map)[i];
    if (path.startsWith(key)) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      return map[key];
    }
  }
};
