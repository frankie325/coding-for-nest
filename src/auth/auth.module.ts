import { Global, Module } from '@nestjs/common';
import { AuthController } from './auth.controller';
import { AuthService } from './auth.service';
import { UserModule } from '../user/user.module';
import { JwtModule } from '@nestjs/jwt';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { ConfigEnum } from '../../config/enum';
import { PassportModule } from '@nestjs/passport';
// import { LocalStrategy } from './local.strategy';
import { JwtStrategy } from './jwt.strategy';
import { CaslAbilityService } from './casl-ability.service';

@Global()
@Module({
  imports: [
    ConfigModule,
    UserModule,
    PassportModule,
    // 异步加载，从configService获取SECRET
    JwtModule.registerAsync({
      imports: [ConfigModule],
      useFactory: (configService: ConfigService) => ({
        secret: configService.get<string>(ConfigEnum.JWT_SECRET),
        signOptions: { expiresIn: '1d' }, // 设置JWT的过期时间为1天
      }),
      inject: [ConfigService],
    }),
  ], //引入UserModule，可以在AuthService中使用UserService
  controllers: [AuthController],
  providers: [AuthService, JwtStrategy, CaslAbilityService],
  exports: [AuthService, CaslAbilityService],
})
export class AuthModule {}
