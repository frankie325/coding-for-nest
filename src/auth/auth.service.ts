import { ForbiddenException, Injectable } from '@nestjs/common';
import { UserService } from '../user/user.service';
import { JwtService } from '@nestjs/jwt';
import { User } from 'src/user/user.entity';
import * as argon2 from 'argon2';
@Injectable()
export class AuthService {
  constructor(
    private userService: UserService,
    private jwtService: JwtService,
  ) {}

  /*
   signAsync 生成jwtWebToken
  
  使用@UseGuards(AuthGuard('jwt'))守卫 => 调用jwtStrategy => 验证前段请求头携带的jwt
  */
  async signin(username: string, password: string) {
    const user = await this.userService.find(username);

    // 如果用户不存在，抛出403 Forbidden异常
    // 这里可以根据实际需求修改异常信息
    // 比如可以抛出一个自定义的异常类，或者使用内置的HttpException类
    if (!user) {
      throw new ForbiddenException('用户不存在，请注册');
    }

    // 使用argon2验证密码
    const isPasswordValid = await argon2.verify(user.password, password);

    // 如果密码不正确，抛出403 Forbidden异常
    if (!isPasswordValid) {
      throw new ForbiddenException('用户名或密码错误');
    }

    const payload = { sub: user.id, username: user.username };
    return {
      access_token: await this.jwtService.signAsync(payload),
      // access_token: 'Bearer ' + (await this.jwtService.signAsync(payload)), 前端要记得加上 Bearer 前缀
    };
  }

  // 用户注册
  async signup(username: string, password: string) {
    // 检查用户名是否已存在
    const existingUser = await this.userService.find(username);
    if (existingUser) {
      throw new ForbiddenException('用户名已存在，请更换用户名');
    }
    // 对密码进行哈希处理
    password = await argon2.hash(password);
    const user = await this.userService.create({
      username,
      password,
    });
    return user;
  }

  async validateUser(
    username: string,
    password: string,
  ): Promise<Partial<User> | null> {
    const user = await this.userService.find(username);
    if (user && user.password === password) {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { password, ...result } = user;
      return result;
    }
    return null;
  }
}
