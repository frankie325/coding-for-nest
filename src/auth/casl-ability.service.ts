import { Injectable } from '@nestjs/common';
import { createMongoAbility, AbilityBuilder } from '@casl/ability';
import { UserService } from '../user/user.service';
import { getEntities } from '../utils/common';

/*
acl存储的其它思路：

acl也可以创建一个单独的表进行存储
*/
@Injectable()
export class CaslAbilityService {
  constructor(private userService: UserService) {}
  // 名字自定义，针对于整个系统的权限
  async forRoot(username: string) {
    const { can, build } = new AbilityBuilder(createMongoAbility);

    // manage和all是casl内部的两个保留字段：表示允许访问所有资源
    // can('manage', 'all');
    const user = await this.userService.find(username);
    // user -> 1:n roles -> 1:n menus：实际需要去重
    user?.roles.forEach((role) => {
      role.menus.forEach((menu) => {
        const actions = menu.acl.split(',');
        for (let i = 0; i < actions.length; i++) {
          const action = actions[i];
          // casl库设置acl策略
          // acl策略create,update,delete,manage,read -> 实体
          can(action, getEntities(menu.path));
        }
      });
    });
    const ability = build({
      detectSubjectType: (object) => object.constructor.name,
    });

    return ability;
  }
}
