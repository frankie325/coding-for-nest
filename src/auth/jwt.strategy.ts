import { ExtractJwt, Strategy } from 'passport-jwt';
import { PassportStrategy } from '@nestjs/passport';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { ConfigEnum } from '../../config/enum';

/*
JWT Strategy:
1.客户端发送带 token 的请求
2.验证 token 有效性
3.解析 token 获取用户信息
4.允许访问受保护的资源

Local Strategy 用于登录时用户，密码的初始认证
而 JWT Strategy 用于后续的会话管理
*/
@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy) {
  constructor(protected configService: ConfigService) {
    const jwtSecret = configService.get<string>(ConfigEnum.JWT_SECRET);
    // console.log('🚀 ~ JwtStrategy ~ jwtSecret:', jwtSecret);

    super({
      // 从请求中提取 JWT 的方法
      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
      ignoreExpiration: false,
      secretOrKey: jwtSecret as string, // 使用秘钥进行签名验证
    });
  }
  // Passport 首先验证 JWT 的签名并解码 JSON 。然后调用我们的 validate() 方法
  validate(payload: { sub: string; username: string }) {
    // console.log('🚀 ~ JwtStrategy ~ payload:', payload);

    // Passport 将基于 validate() 方法的返回值构建一个user 对象，并将其作为属性附加到请求对象上
    return { id: payload.sub, username: payload.username };
  }
}
