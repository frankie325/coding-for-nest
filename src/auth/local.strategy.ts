import { Strategy } from 'passport-local';
import { PassportStrategy } from '@nestjs/passport';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { AuthService } from './auth.service';

/*
passport本地策略：用于初始登录，需要在请求体中传递用户名和密码，验证成功后通常返回 JWT token

Local Strategy:
1.客户端发送用户名密码
2.验证用户名密码
3.生成 JWT token
4.返回 token 给客户端
*/

@Injectable()
export class LocalStrategy extends PassportStrategy(Strategy) {
  constructor(private readonly authService: AuthService) {
    super();
  }

  async validate(username: string, password: string): Promise<any> {
    const user = await this.authService.validateUser(username, password);
    console.log('🚀 ~ LocalStrategy ~ validate ~ user:', user);

    if (!user) {
      throw new UnauthorizedException();
    }
    return user;
  }
}
