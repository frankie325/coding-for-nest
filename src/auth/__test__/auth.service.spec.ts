import { Test, TestingModule } from '@nestjs/testing';
import { User } from '../../user/user.entity';
import { UserService } from '../../user/user.service';
import { AuthService } from '../auth.service';
import { JwtService, JwtSignOptions } from '@nestjs/jwt';
import { ForbiddenException } from '@nestjs/common';

describe('AuthService（登陆认证模块-服务）', () => {
  let authService: AuthService;
  let userService: Partial<UserService>;
  let jwt: Partial<JwtService>;
  let userArr: User[] = []; //模拟数据库，单元测试并不是真的链接到数据库，存储用户信息缓存，判断用户是否注册过

  const mockUser = {
    username: 'test',
    password: '123456',
  };

  beforeEach(async () => {
    userService = {
      find: (username: string) => {
        const user = userArr.find((user) => user.username === username);
        return Promise.resolve(user || null);
      },
      create: async (user: User) => {
        const tempUser = new User();
        tempUser.id = Math.floor(Math.random() * 1000);
        tempUser.username = user.username;
        tempUser.password = user.password;
        userArr.push(tempUser);
        return Promise.resolve(tempUser);
      },
    };
    jwt = {
      signAsync(payload: string | Buffer | object, options?: JwtSignOptions) {
        return Promise.resolve('token');
      },
    };

    const module: TestingModule = await Test.createTestingModule({
      // 提供AuthService中相关的依赖
      providers: [
        AuthService,
        {
          provide: UserService,
          useValue: userService,
        },
        {
          provide: JwtService,
          useValue: jwt,
        },
      ],
    }).compile();

    authService = module.get<AuthService>(AuthService);
  });

  //   测试完之后，清空缓存
  afterEach(() => {
    userArr = [];
  });

  it('鉴权-初始化-实例化', () => {
    // 断言authService已经定义
    expect(authService).toBeDefined();
  });

  it('用户初次注册', async () => {
    const user = await authService.signup(mockUser.username, mockUser.password);
    expect(user).toBeDefined();
    expect(user.username).toBe(mockUser.username);
  });

  it('用户使用相同的用户名注册', async () => {
    await authService.signup(mockUser.username, mockUser.password);
    await expect(
      authService.signup(mockUser.username, mockUser.password),
    ).rejects.toThrowErrorMatchingSnapshot(); //生成一个快照测试

    // 如果测试抛出的错误与真实的代码抛出的错误不一致，则报错。说明真实的代码写错了
    // await expect(
    //   authService.signup(mockUser.username, mockUser.password),
    // ).rejects.toThrow(new ForbiddenException('用户名已存在，请更换用户名1'));
    await expect(
      authService.signup(mockUser.username, mockUser.password),
    ).rejects.toThrow(new ForbiddenException('用户名已存在，请更换用户名'));
    /*
    快照测试：
        第一次运行测试时，Jest 会捕获错误信息并创建快照文件
        后续运行时，将新的错误信息与存储的快照进行比较
        快照文件通常保存在 __snapshots__ 目录下
    */
  });

  it('用户登录', async () => {
    // 注册新用户
    await authService.signup(mockUser.username, mockUser.password);
    // 登陆
    const res = await authService.signin(mockUser.username, mockUser.password);
    expect(res.access_token).toBe('token');
  });

  it('用户登录-密码错误', async () => {
    // 注册新用户
    await authService.signup(mockUser.username, mockUser.password);
    // 登陆
    await expect(
      authService.signin(mockUser.username, '111111'),
    ).rejects.toThrow(new ForbiddenException('用户名或密码错误'));
  });

  it('用户登录-用户名不存在', async () => {
    // 注册新用户
    await authService.signup(mockUser.username, mockUser.password);
    // 登陆
    await expect(authService.signin('test222', '111111')).rejects.toThrow(
      new ForbiddenException('用户不存在，请注册'),
    );
  });
});
