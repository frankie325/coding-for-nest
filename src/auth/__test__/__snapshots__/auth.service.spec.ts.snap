// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService（登陆认证模块-服务） 用户使用相同的用户名注册 1`] = `"用户名已存在，请更换用户名"`;
